pkg/infra/mock/metadata_mock.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"log"
+	"net/http"
+	"strings"
+)
+
+const (
+	awsTerminationPattern   = "/latest/meta-data/spot/termination-time"
+	gcpTerminationPattern   = "/computeMetadata/v1/instance/preempted"
+	azureTerminationPattern = "/metadata/scheduledevents"
+)
+
+func main() {
+	provider := flag.String("provider", "", "Cloud Provider metadata service to mock (One of AWS, Azure or GCP)")
+	listenAddr := flag.String("listen-addr", "0.0.0.0:80", "Address on which metadata mock service should listen")
+	flag.Parse()
+
+	var handler http.Handler
+	switch {
+	case strings.EqualFold(*provider, "aws"):
+		handler = awsMetadataMockHandler()
+	case strings.EqualFold(*provider, "azure"):
+		handler = azureMetadataMockHandler()
+	case strings.EqualFold(*provider, "gcp"):
+		handler = gcpMetadataMockHandler()
+	default:
+		log.Fatal("--provider must be one of: aws, azure or gcp")
+	}
+
+	log.Printf("Starting mock metadata service for provider: %s", *provider)
+	if err := http.ListenAndServe(*listenAddr, logRequests(handler)); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// AWS instances expect an OK response to indicate that the instance has been scheduled for termination
+func awsMetadataMockHandler() http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(awsTerminationPattern, func(rw http.ResponseWriter, req *http.Request) {
+		rw.WriteHeader(http.StatusOK)
+	})
+
+	return mux
+}
+
+// Azure instances expect an OK response with a Json body containing a schedule peremption event to
+// indicate that the instance has been scheduled for termination.
+// Requires the header "Metadata: true".
+// Requires a query with the api version: eg `?api-version=2019-08-01`.
+func azureMetadataMockHandler() http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(azureTerminationPattern, func(rw http.ResponseWriter, req *http.Request) {
+		if req.Header.Get("Metadata") != "true" {
+			// Require the "Metadata" header to be correct
+			rw.WriteHeader(http.StatusBadRequest)
+			rw.Write([]byte("400 Bad Request"))
+			return
+		}
+
+		if req.URL.Query().Get("api-version") != "2019-08-01" {
```

Comment:
given that this api version value will change at some point in the future, i think it would be nice for visibility to pull this out as a top level var or const. just a suggestion though, not a blocker for me.

Reply:
Yep, seems good. We may want to transition it at some point in the future as well, might need to make it one of a set of supported values, but I'll leave that for when we need to do it